### PR TITLE
Minos push (#589)

### PR DESCRIPTION
**Pull request by thesuzerain**
- Adds handling for Minos backend and new authentication system, for simultaneous use with the current version of Minos
- Adds APT generation and use endpoint

Currently incomplete.

Fixes MOD-265
Fixes MOD-399
Fixes MOD-269
Fixes MOD-264
Fixes MOD-269

### CODE REVIEW

**Review by Geometrically — CHANGES_REQUESTED** [review]
this is my old review

File: src/database/models/ids.rs [hunk 1]
```diff
@@ -191,6 +203,10 @@ pub struct ThreadId(pub i64);
 #[sqlx(transparent)]
 pub struct ThreadMessageId(pub i64);
 
+#[derive(Copy, Clone, Debug, Type)]
+#[sqlx(transparent)]
+pub struct PatTokenPart(pub i64);
+
```
[the remark this answers is not in the document]
Comment (thesuzerain) [reply src/database/models/ids.rs:209]: Leftover. Was originally going to store a PAT as two base62 encoded i64s as i128 cannot be stored by sqlx. Eventually simple made it a string

**Review by Geometrically — CHANGES_REQUESTED** [review]

File: src/routes/v2/admin.rs [hunk 2]
```diff
@@ -1,9 +1,12 @@
+use crate::database::models::user_item;
 use crate::models::ids::ProjectId;
 use crate::models::projects::MonetizationStatus;
+use crate::models::users::User;
 use crate::routes::ApiError;
+use crate::util::auth::{link_or_insert_new_user, MinosNewUser};
 use crate::util::guards::admin_key_guard;
-use crate::DownloadQueue;
-use actix_web::{patch, post, web, HttpResponse};
+use crate::{models, DownloadQueue};
```
Comment (Geometrically) [inline src/routes/v2/admin.rs:8]: clippy

File: src/routes/v2/admin.rs [hunk 3]
```diff
@@ -16,10 +19,43 @@ pub fn config(cfg: &mut web::ServiceConfig) {
...
+    println!("Here :)");
+    let github_id = github_id.into_inner();
+    println!("Github id: {}", github_id);
+    let mut transaction = client.begin().await?;
+    let user = user_item::User::get_from_github_id(github_id as u64, &mut *transaction).await?;
+    let user: Option<User> = user.map(|u| u.into());
+    println!("User found");
+    transaction.commit().await?;
+    Ok(HttpResponse::Ok().json(user))
+}
```
Comment (Geometrically) [inline src/routes/v2/admin.rs:57]: print statements

Comment (Geometrically) [reply src/routes/v2/admin.rs:57]: also why is there a transaction for get legacy account? can't it just use the pool directly

File: src/util/auth.rs [hunk 4]
```diff
@@ -18,85 +27,225 @@ pub enum AuthenticationError {
...
+        (_, Some(cookies)) => {
+            minos_user = get_minos_user_from_cookies(
+                cookies
+                    .to_str()
+                    .map_err(|_| AuthenticationError::InvalidCredentials)?,
+            )
+            .await?;
+            db_user = models::User::get_from_minos_kratos_id(minos_user.id.clone(), executor)
+                .await?
+                .ok_or_else(|| AuthenticationError::InvalidCredentials)?;
```
Comment (Geometrically) [inline src/util/auth.rs:198]: instead of throwing errors when each auth method fails, shouldn't it try the next one?

**Review by Geometrically — CHANGES_REQUESTED** [review]

File: src/util/auth.rs [hunk 5]
```diff
@@ -18,85 +27,225 @@ pub enum AuthenticationError {
...
-        .get("Authorization")
-        .ok_or(AuthenticationError::InvalidCredentials)?
-        .to_str()
-        .map_err(|_| AuthenticationError::InvalidCredentials)?;
-
-    get_user_from_token(token, executor).await
+    if token.starts_with("Bearer ") {
+        let token: &str = token.trim_start_matches("Bearer ");
+
+        // Tokens beginning with Ory are considered to be Kratos tokens (extracted cookies) and forwarded to Minos
```
[the remark this answers is not in the document]
Comment (thesuzerain) [reply src/util/auth.rs:239]: Small unfinished thing I forgot about. Added functionality here, will add matching functionality to Minos after.

**Review by Geometrically — APPROVED** [review]

**Pull request merged**